Updating the default cipher-suite lists for the 9.x release.

shinrich opened the pull request:
Based on comments from issue #5389.  Starting the ball rolling to figure out the appropriate default cipher_suite lists for the 9.x release.

Added defaults for the TLS1.3 client and server settings, selecting CHACHA first for performance reasons.

For the server cipher_suites, Added the CHACHA20-POLY and preferred some shorter AES versions over longer ones for performances reasons.

Added defaults for the client side.  In general the client side should be more generous because you don't control the servers to talk to particularly in the forward proxy case.  Based on feedback from an internal security person, I got this rather large list by running 
```
openssl ciphers 'ALL:!aNULL:!aDH:!aECDH:!aPSK:!aSRP:!eNULL:!kSRP:!SSLv2:!SSLv3:!TLSv1:!TLSv1.1:@STRENGTH'
```
To get all sensible ciphers for TLS1.2 and beyond.  Open for discussion on this one.  We should definitely have something better than NULL (which allows for all ciphers).

zwoop commented:
I think CHACHA20 is the slowest of them all :).

maskit commented on mgmt/RecordsConfig.cc:
```diff
@@ -1174,9 +1175,9 @@ static const RecordElement RecordsConfig[] =
   // The default value (nullptr) means the default value of TLS stack will be used.
   // - e.g. OpenSSL : "TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256"
-  {RECT_CONFIG, "proxy.config.ssl.server.TLSv1_3.cipher_suites", RECD_STRING, nullptr, RECU_RESTART_TS, RR_NULL, RECC_NULL, nullptr, RECA_NULL}
+  {RECT_CONFIG, "proxy.config.ssl.server.TLSv1_3.cipher_suites", RECD_STRING, "TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384", RECU_RESTART_TS, RR_NULL, RECC_NULL, nullptr, RECA_NULL}
```
This doesn't match with the comment above.

shinrich commented:
As a stream cipher, CHACHA20-POLY is faster on a general CPU (like a mobile phone).  With hardware assist AES may be faster.

shinrich commented:
Got rid of the old comment.

randall commented:
[approve ci autest]

bryancall commented on mgmt/RecordsConfig.cc:
```diff
@@ -1172,11 +1173,9 @@ static const RecordElement RecordsConfig[] =
-  // The default value (nullptr) means the default value of TLS stack will be used.
-  // - e.g. OpenSSL : "TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256"
-  {RECT_CONFIG, "proxy.config.ssl.server.TLSv1_3.cipher_suites", RECD_STRING, nullptr, RECU_RESTART_TS, RR_NULL, RECC_NULL, nullptr, RECA_NULL}
+  {RECT_CONFIG, "proxy.config.ssl.server.TLSv1_3.cipher_suites", RECD_STRING, "TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384", RECU_RESTART_TS, RR_NULL, RECC_NULL, nullptr, RECA_NULL}
```
It doesn't make sense to prefer CHACHA on the server side since most servers will have AES support build into their CPUs and will be faster.

bryancall replied:
Benchmarked on my Mac and AES is ~6 times faster:
```
15:30:36 C02XF1ULJG5J:(std_array)~/dev/apache/trafficserver$ openssl speed aes-128-gcm chacha20-poly1305
Doing aes-128 gcm for 3s on 16 size blocks: 40891550 aes-128 gcm's in 2.98s
Doing aes-128 gcm for 3s on 64 size blocks: 33355414 aes-128 gcm's in 2.97s
Doing aes-128 gcm for 3s on 256 size blocks: 19007676 aes-128 gcm's in 2.99s
Doing aes-128 gcm for 3s on 1024 size blocks: 6378002 aes-128 gcm's in 2.99s
Doing aes-128 gcm for 3s on 8192 size blocks: 803093 aes-128 gcm's in 2.97s
Doing chacha20 poly1305 for 3s on 16 size blocks: 6373621 chacha20 poly1305's in 2.98s
Doing chacha20 poly1305 for 3s on 64 size blocks: 6446820 chacha20 poly1305's in 2.98s
Doing chacha20 poly1305 for 3s on 256 size blocks: 2987367 chacha20 poly1305's in 2.97s
Doing chacha20 poly1305 for 3s on 1024 size blocks: 1019862 chacha20 poly1305's in 2.98s
Doing chacha20 poly1305 for 3s on 8192 size blocks: 143181 chacha20 poly1305's in 3.00s
```

c-taylor commented:
Some other implementations honour client preference only if ChaCha20 is at the head as a loose indication that the client has no AES offload, allowing implementors to choose whether to take the hit on the server.

I have recently tested many generations of CPU (up to at least 5 years) and ChaCha is always second place to GCM ciphers, these all had offload capability. Interestingly ChaCha did outpace AES CBC cipher suites on larger block sizes (8/16k) even with offload, but this is only a TLSv1.2 concern and possibly an edge case. i.e. ChaCha20 support but no (AES GCM||AES offload).

shinrich commented:
Thanks for all the feedback.  Updated the defaults based on feedback.  The guidance for this version is 

PFS ciphers first.  Within the PFS ciphers, GCM and CCM AES ciphers.  Then CHACHA, then CBC ciphers.

shinrich commented:
[approve ci autest]

shinrich commented:
One more update to remove the cipher_suite entry from records.config.default.in since it was just tracking the traffic_server default.

maskit commented on mgmt/RecordsConfig.cc:
```diff
@@ -1172,11 +1173,9 @@ static const RecordElement RecordsConfig[] =
-  // The default value (nullptr) means the default value of TLS stack will be used.
-  // - e.g. OpenSSL : "TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256"
-  {RECT_CONFIG, "proxy.config.ssl.server.TLSv1_3.cipher_suites", RECD_STRING, nullptr, RECU_RESTART_TS, RR_NULL, RECC_NULL, nullptr, RECA_NULL}
+  {RECT_CONFIG, "proxy.config.ssl.server.TLSv1_3.cipher_suites", RECD_STRING, "TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256", RECU_RESTART_TS, RR_NULL, RECC_NULL, nullptr, RECA_NULL}
```
We prefer TLS_AES_256_GCM_SHA384 rather than TLS_AES_128_GCM_SHA256?
I'm not an expert here, but it might be better to use the default value that TLS stack have, like the removed comment says?

@masaori335 What do you think?

masaori335 replied:
I'm not sure we need to tweak default settings, because what is best is depends on use case and there'r no TLSv1.3 cipher suites which is too weak and we should remove from default settings.

shinrich replied:
We run with AES128 in preference to AES256 for performance.  That is also the ordering for the non-TLS ciphers.

I am ok with going with AES256 first since this is just a default . In my opinion either ordering would be safe enough, but we should be consistent with the non-TLS1.3 list as well.  Going with AES256 first was the default for 8.x, and we had a rather lengthy discussion about the defaults for 7.x or 8.x.  Will update PR for that position.

shinrich commented:
Updated the TLSv1_3.cipher_suites to explicitly enumerate the openssl default ordering.  I think it is worthwhile to be explicit even if it is the openssl -1.1.1 default.

Took another pass on the proxy.config.ssl.server.cipher_suite.  Here is my dimension outline and breakdown rationale for this list.  First, no ciphers less that TLSv1.2, since the default configuration has protocols less than TLSv1.2 turned off.   This gets rid of triple DES and the SHA1 algorithms.

1.  Prefer PFS over non-PFS
2. Key exchange: prefer ECDHE over DHE. 
3. Authentication: prefer ECDSA over RSA
4. Mode: GCM, CCM, CCM8, CBC
5. Encryption algorithm: AES256, AES128.  Insert the CHACHA20 stream cipher before AES in CBC mode.

I have removed the common "!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA" suffix.  It doesn't help for this specific sequence.  Perhaps worthwhile to put some catch-all at the end to help if people start messing with the default option?

shinrich commented:
[approve ci autest]

shinrich commented:
[ci approve autest]

shinrich commented:
[approve ci autest]

shinrich commented:
[approve ci clang-analyzer]

maskit commented:
> Updated the TLSv1_3.cipher_suites to explicitly enumerate the openssl default ordering. I think it is worthwhile to be explicit even if it is the openssl -1.1.1 default.

I'm ok with this, but if we do so, we'll need to update the default value when OpenSSL update its default cipher suites and ordering. New cipher suites will never be available until we update our default value or users update their configuration to use it.

zwoop commented:
Cherry-picked to v9.0.x branch.
